src/GraphQL.DataLoader/DataLoaderDocumentListener.cs

```diff
@@ -26,15 +26,15 @@ public Task AfterValidationAsync(IExecutionContext context, IValidationResult va
     /// <inheritdoc/>
     public Task BeforeExecutionAsync(IExecutionContext context)
     {
-        _accessor.Context ??= new();
```

Comment:
The tests were failing because this did a null check before setting to a new instance. That no longer works now that we throw if it's `null`.

The obvious fix was to add a function named `Initialize` that could bypass the getter, accessing the `AsyncLocal` directly and setting it to a new instance if it was `null`. I then also decided to remove the setter altogether and introduce a function to reset the context back to `null`.

I'm not sure the setter makes sense anymore since you can't do a conditional set based on whether or not it is `null`. This is why I went ahead and removed it and replace it with a `Reset()` function to set it back to `null`.

As I write this, I realize it's a larger breaking change because consumers can no longer set the context to an arbitrary value. So maybe a) the setter should remain or b) `Reset` should take in a `DataLoaderContext?` parameter so allow consumers to set it to whatever they want.

Reply:
Maybe we should reconsider simply changing the NRT attributes. 

Also consider that this interface is modeled after `IHttpContextAccessor` and we should review/consider their implementation. 

https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.ihttpcontextaccessor.httpcontext?view=aspnetcore-7.0#microsoft-aspnetcore-http-ihttpcontextaccessor-httpcontext

Reply:
Another option is to leave existing property as is and introduce a new property like RequiredContext (similar in vein to GetRequiredService). Sorta wish it was an extension method though

Reply:
Regarding it being modeled against `IHttpContextAccessor`, I think a good question to ask is what are valid use cases for it. In one [documentation example](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/use-http-context?view=aspnetcore-7.0#httpcontext-isnt-thread-safe) the `IHttpContextAccessor` is used in a `HttpClient` delegating handler, where depending on when the `HttpClient` is called, could result in being outside of the request flow and having the `HttpContext` nullable.

Is it considered valid usage of the `DataLoaderContext` for it to be accessed outside of document execution? If it is, the property should stay nullable and something like `RequiredContext`/`GetRequiredContext` could be introduced. If it's not valid to use it outside of document execution, then I think it's better to have it modeled against the valid use case (non-null).

----------------

Another interesting discussion on nullability is [here](https://github.com/dotnet/aspnetcore/pull/22928#discussion_r440513476), when `ISession` is changed to be non-null, even though technically custom implementations of `HttpContext` could cause it to return a non-null instance. I like the comment that there needs to be pragmatism applied to nullability attributes and focus on the 99% use case, and I think their [implementation in `DefaultHttpContext`](https://github.com/dotnet/aspnetcore/blob/744cb017a7d7d14a5e25076b9a9a8d3e4702f5d8/src/Http/Http/src/DefaultHttpContext.cs#L198) is similar to what I'm attempting here in the previous commit.

So another possible option is to revert the last commit that adds `Reset` and `Initialize`, which effectively disallows the `??=` operator. But this is not really different from what was done with `DefaultHttpContext.Session`. This makes the assumption that changing the way `DataLoaderDocumentListener` sets the context is safe:

```csharp
// before
_accessor.Context ??= new();

// after
_accessor.Context = new();
```
It seems like it's safe to skip the null check and always set it (I assume each `IDocumentExecutionListener` is only executed once). If a custom listener sets their own, it would require that they ensure `DataLoaderDocumentListener` is not added.

Reply:
Well, to think about it a bit, having `IHttpContextAccessor` return a nullable context allows code to operate differently based on whether it was null, and is useful for logging and similar scenarios.  But pulling the `IDataLoaderContext` is really only useful if you're going to use the data loader, which implies returning an `IDataLoaderResult<T>` from the method.  Sure it's theoretically possible to detect that the context is null and immediately retrieve the values, but you'd still need to wrap the values in an `IDataLoaderResult<T>`.  As such, the only practical use would be for tests, and it's much simpler to mock the data loader context in that scenario.

So I'd have to say that it's not really 'valid usage' to use `IDataLoaderContextAccessor` without the listener in use.

Agree?

Reply:
Let's go with your suggestion above, reverting the last commit and changing this:

```cs
// before
_accessor.Context ??= new();

// after
_accessor.Context = new();
```

If someone wants to be able to read `null` values from the context, they can write their own implementation of `DataLoaderContextAccessor` which allows null values or provides `Initialize` and `Reset` methods or similar.

Reply:
(Which is how the PR is now.)

Reply:
> So I'd have to say that it's not really 'valid usage' to use IDataLoaderContextAccessor without the listener in use.

I very much agree. I was thinking about it yesterday and came to the same conclusion about the data loader result.